### ISSUE TITLE
Need help getting started

### ISSUE THREAD

**Issue by tahmidkhan** [chandlerprall/DatabindExtension#19]
Hello, thanks for this library. Really seems like it could be very helpful. Unfortunately, I can't get started following the guide. I have the following code on my JS file:

```javascript
var nunjucks = require('nunjucks'),
    databindExtension = require('nunjucks-databind')

var nunjucksEnv = nunjucks.configure({ autoescape: true })
nunjucksEnv.addExtension('BindExtension', new DatabindExtension({}))

function render(template, context) {
    return nunjucks.renderString(template, context)
}

console.log(render('Hello {{ username }}', { username: 'admin'}))
```

When I try running this file, I get this error: 

```
ReferenceError: DatabindExtension is not defined
    at Object.<anonymous> (/Users/<username>/<project>/test.js:5:47)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
    at internal/main/run_main_module.js:17:47
```

What exactly am I doing wrong here?

**Comment by ultimate-tester** [comment]
Hi there, thanks for using this extension.

The error here is that you are trying to create an instance of an undefined class. This is because the name of the variable you created when requiring the module is different from the spot where you're trying to initialize the object. They should be the same (and this is case sensitive).

**Comment by tahmidkhan** [comment]
@ultimate-tester Thanks for the quick reply! I tried doing this:

```javascript
var nunjucks = require('nunjucks'),
    DatabindExtension = require('nunjucks-databind')

var nunjucksEnv = nunjucks.configure({ autoescape: true })
nunjucksEnv.addExtension('BindExtension', new DatabindExtension({}));

function render(template, context) {
    return nunjucks.renderString(template, context)
}

console.log(render('Hello {{ username }}', { username: 'admin'}))
```

Now I get this error:

```
return typeof window['Proxy'] === 'function';
      ^

ReferenceError: window is not defined
    at Function.detectProxyFeature (/Users/<username>/<project>/node_modules/nunjucks-databind/dist/DatabindExtension.js:1040:7)
    at Function.determineUpdateMode (/Users/<username>/<project>/node_modules/nunjucks-databind/dist/DatabindExtension.js:1029:29)
    at new DatabindExtension (/Users/<username>/<project>/node_modules/nunjucks-databind/dist/DatabindExtension.js:913:43)
    at Object.<anonymous> (/Users/<username>/<project>/test.js:5:43)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
    at internal/main/run_main_module.js:17:47
```

**Comment by ultimate-tester** [comment]
Alright, then that error is solved.
From the next error I conclude that you're using Nunjucks inside NodeJS. Nunjucks itself is a library that works both on server (i.e. NodeJS) or on clients (i.e. browsers). This extension in particular, only works on the client side.

If you really want synchronization of context between server and client you need to use websockets (or a library that uses this). This way you will have a real-time communication and can sync the context.

If you want to make this module work, render your views on the client. Check the guide on the Nunjucks website.

**Comment by tahmidkhan** [comment]
@ultimate-tester Hmm, I think I understand what you mean. Just had one other question. Provided I do get this to work on client side, **is there a dedicated function to update an attribute in the context**? In the example scenario I'm thinking, I would like to attach an event listener to an input and use the value from that input to update the context so that the changes are reflected on my template as well. 

Do I just call `updateBindings([force])` every time I want the template to update? Let's say I create the context as follows:

```Javascript
var context = databindExtension.createContext({
    displayName: 'person',
    items: ['Test1', 'Test2']
});
```

How would I then update the `displayName` attribute to `"admin"`? That's basically what I'm asking.

**Comment by ultimate-tester** [comment]
No worries, that's the beauty of this module. You just have to setup the context in a specific way like demonstrated in the readme:

`var context = databindExtension.createContext({
    displayName: 'person',
    items: ['Test1', 'Test2']
});`

Now, you can do the following to manipulate the context and it will immediately reflect on your template:

`context.displayName = "John";`

No special method, just regular property manipulation. Any operator should work fine as well as array methods (e.g. push) etc.

**Comment by tahmidkhan** [comment]
That's aces! Thanks a lot. I'll be closing this issue then.

**Comment by ultimate-tester** [comment]
Also, check the examples folder. It does exactly what you are describing.

**Comment by tahmidkhan** [comment]
@ultimate-tester Sorry to re-open this again. But I created this template and ran it on the browser:

```html
<!DOCTYPE html>
<html>
<head>
<title>Nunjucks Databinding Test</title>
</head>

<body>
	<div id="content"></div>

	<!-- Nunjucks -->
	<script src="https://cdn.jsdelivr.net/npm/nunjucks@3.2.2/browser/nunjucks.min.js"></script>

	<!-- Nunjucks Databinding -->
	<script type="text/javascript" src="DatabindExtension.js"></script>

	<script type="text/javascript">
		var content = ""
		var contentContainer = document.getElementById("content")

		var contentString = `
		<h1>Hello {{ username }}</h1>

		<input type="text" value="{{ username }}" id="username" />
		`

		var nunjucksEnv = nunjucks.configure({ autoescape: true })
		var databindExtension = new DatabindExtension()
		nunjucksEnv.addExtension('BindExtension', databindExtension)
		var context = databindExtension.createContext({ username: 'admin' })

		content = nunjucks.renderString(contentString, context)
		contentContainer.innerHTML = content

		document.getElementById("username").addEventListener("input", function(event) {
			console.log(event.target.value)
			context.username = event.target.value
		})
	</script>

</body>
</html>
```
The first load works, which means that the library is being imported properly, and the context is also being set. However, when I update the input box, the context doesn't seem to change. Am I doing something wrong?

**Comment by ultimate-tester** [comment]
Only thing I can ask at this time is whether the console.log is triggering? If not try the keydown event.

If that's not it, can you tell me which browser you tried? I'll investigate over the weekend.

**Comment by tahmidkhan** [comment]
@ultimate-tester To answer your questions:
- Yep, the console logging works and is outputting the correct value
- Keydown doesn't work either. Again, the console logging works
- My browser is latest Chrome on Mac OSX

**Comment by tahmidkhan** [comment]
Ok so after doing some digging, it seems like this library is meant to work only with the `Template` class of objects from the Nunjucks library. For that reason, if anyone wants to use a string (instead of an actual template like all the examples), they would need to do the following:

```javascript
var myTemplate = new nunjucks.Template(templateString, nunjucksEnv)
myTemplate.render(data, function(error, result) {
   if (!!error) {...}
   ...
});
```

@ultimate-tester Please feel free to close this issue if this library is indeed meant to be used only with the `Template` class.

**Comment by ultimate-tester** [comment]
Hi again, I had this on the backlog for investigation today but I think your answer leads me to it:

You are loading the extension into the nunjucksEnv. Therefore, you wil also need to use the render methods of the environment. Your earlier code snippet uses the global nunjucks variable in which the extension is not loaded. Your code of today shows that you are using the environment. That's why it's working.

**Tl;dr:** use nunjucksEnv.renderString() or nunjucksEnv.render()

If that's not it, then you found a bug

**Comment by tahmidkhan** [comment]
Nvm, you are right, this actually solves it. The environment needs to be used to call the render function. Thank you!
